lib/instrumentation/index.js

```diff
@@ -358,35 +361,43 @@ Instrumentation.prototype._startHook = function () {
 
   this._agent.logger.debug('adding hooks to Node.js module loader');
 
-  this._ritmHook = new RitmHook(this._patches.keys, function (
-    exports,
-    name,
-    basedir,
-  ) {
-    const enabled = self._isModuleEnabled(name);
-    var pkg, version;
-
-    const isHandlingLambda =
-      self._lambdaHandlerInfo && self._lambdaHandlerInfo.module === name;
-
-    if (!isHandlingLambda && basedir) {
-      pkg = path.join(basedir, 'package.json');
-      try {
-        version = JSON.parse(fs.readFileSync(pkg)).version;
-      } catch (e) {
-        self._agent.logger.debug(
-          'could not shim %s module: %s',
-          name,
-          e.message,
-        );
-        return exports;
+  this._ritmHook = new RitmHook(
+    this._patches.keys,
+    // XXX: found this was not enabled so our redis instrumentations
+    // may not be working as for now
+    { internals: true },
```

Comment:
Note for reviewer: check the comment and give feedback please. Was okay to pass the new option? what about other instrumentations.

Reply:
`internals: true` isn't needed if the sub-module path, e.g. `@redis/client/dist/lib/client`, is an entry in the first argument to `RitmHook` -- i.e. if it is on the explicit list of modules to hook into. If I add these debug-prints to RITM:

```diff
--- /Users/trentm/el/require-in-the-middle/index.js	2023-07-13 15:32:26
+++ ./node_modules/require-in-the-middle/index.js	2023-10-03 09:26:14
@@ -126,6 +126,7 @@
   const self = this
   const patching = new Set()
   const internals = options ? options.internals === true : false
+  console.log('XXX ritm: internals=%s', internals);
   const hasWhitelist = Array.isArray(modules)

   debug('registering require hook')
@@ -204,6 +205,7 @@
       }
       moduleName = filename
     } else if (hasWhitelist === true && modules.includes(filename)) {
+      console.log('XXX ritm abspath in allowlist: hasWhitelist=%s modules.includes(filename)=%s filename=%s', hasWhitelist, modules.includes(filename), filename);
       // whitelist includes the absolute path to the file including extension
       const parsedPath = path.parse(filename)
       moduleName = parsedPath.name
@@ -227,6 +229,7 @@
       if (hasWhitelist === true && modules.includes(moduleName) === false) {
         if (modules.includes(fullModuleName) === false) return exports // abort if module name isn't on whitelist

+        console.log('XXX ritm allowlisted sub-module: moduleName=%s fullModuleName=%s filename=%s', moduleName, fullModuleName, filename)
         // if we get to this point, it means that we're requiring a whitelisted sub-module
         moduleName = fullModuleName
       } else {
diff --git a/lib/instrumentation/index.js b/lib/instrumentation/index.js
index b90152f7..29e4bbb7 100644
--- a/lib/instrumentation/index.js
+++ b/lib/instrumentation/index.js
@@ -363,6 +363,7 @@ Instrumentation.prototype._startHook = function () {
     name,
     basedir,
   ) {
+    console.log('XXX RitmHook called:', name, basedir);
     const enabled = self._isModuleEnabled(name);
     var pkg, version;

```

and run examples/trace-redis.js:

```
% node examples/trace-redis.js
{"log.level":"info","@timestamp":"2023-10-03T16:30:11.754Z","log":{"logger":"elastic-apm-node"},"agentVersion":"4.0.0","env":{"pid":19696,"proctitle":"node","os":"darwin 22.6.0","arch":"x64","host":"pink.local","timezone":"UTC-0700","runtime":"Node.js v16.20.2"},"config":{"serviceName":{"source":"start","value":"example-trace-redis4","commonName":"service_name"},"serviceVersion":{"source":"default","value":"4.0.0","commonName":"service_version"},"serverUrl":{"source":"default","value":"http://127.0.0.1:8200/","commonName":"server_url"},"logLevel":{"source":"default","value":"info","commonName":"log_level"},"spanCompressionEnabled":{"source":"start","value":false}},"activationMethod":"require","ecs":{"version":"1.6.0"},"message":"Elastic APM Node.js Agent v4.0.0"}
XXX ritm: internals=false
XXX ritm allowlisted sub-module: moduleName=@redis/client fullModuleName=@redis/client/dist/lib/client/commands-queue filename=/Users/trentm/el/apm-agent-nodejs/node_modules/@redis/client/dist/lib/client/commands-queue.js
XXX RitmHook called: @redis/client/dist/lib/client/commands-queue /Users/trentm/el/apm-agent-nodejs/node_modules/@redis/client
XXX RitmHook called: generic-pool /Users/trentm/el/apm-agent-nodejs/node_modules/generic-pool
XXX ritm allowlisted sub-module: moduleName=@redis/client fullModuleName=@redis/client/dist/lib/client filename=/Users/trentm/el/apm-agent-nodejs/node_modules/@redis/client/dist/lib/client/index.js
XXX RitmHook called: @redis/client/dist/lib/client /Users/trentm/el/apm-agent-nodejs/node_modules/@redis/client
XXX RitmHook called: redis /Users/trentm/el/apm-agent-nodejs/node_modules/redis
PING res:  PONG
SET res:  OK
GET res:  bar
MULTI res:  [ 'OK', 'eggs' ]
```

The RITM `opts.internals` value is only used later for internal module paths that are *not* on the explicitly set of modules to hook.

Reply:
I did not get those logs for `mongodb` and I think I know why. The last log inside `./node_modules/require-in-the-middle/index.js` is inside a conditional

```js
if (hasWhitelist === true && modules.includes(moduleName) === false) {
    if (modules.includes(fullModuleName) === false) return exports // abort if module name isn't on whitelist

    console.log('XXX ritm allowlisted sub-module: moduleName=%s fullModuleName=%s filename=%s', moduleName, fullModuleName, filename)
    // more code here ...     
}
```

In my case
- moduleName === 'mongodb'
- fullModuleName === 'mongodb/lib/cmap/connection_pool'

The expression allows to pass sub-modules only if the root module is not in the whitelist. I this case the agent is instrumenting the main module and one sub-module. I think it works on `@redis` because its instrumention only sub-modules.

I see two options here. Let me know what you think
- change the instrumentation and go for sub-modules (need to check if feasible)
- make use of the `internals` option.

Reply:
> * make use of the `internals` option.

I'd prefer to avoid this. It will mean that we will be calling our `onRequire` hook function for every require'd file under each of the modules we instrument (every lib file in ioredis, in express, in fastify, etc.). There *may* be a perf impact of this.

---

Option 3: Remove that limitation in RITM.

I think this is a bug in RITM that you cannot do this.  Having 'mongodb' on the list of modules to patch shouldn't mean that you cannot patch `mongodb/lib/cmap/connection_pool`.  I'll have to stare at this longer... the if/else's in RITM make my eyes swim a bit.

Reply:
1. That's okay. There is no need to rush in a fix there and we still have to check what's the impact since no one hasn't noticed.
2. I've done the changes to have sub-modules with the `.js` extension. I've noticed our Hook is reading the `package.json` file each time is called so enabling internals will do many reads. To avoid this I made use of a `Map` to cache versions.

Reply:
This [PR](https://github.com/elastic/apm-agent-nodejs/pull/3658) solves our use case for instrumenting module and sub-module at the same time. which is what we need for `mongodb`.

We still have to look at RITM but since the issue is only reported by us I think we have no rush.